### ISSUE TITLE
403 completely breaks program

### ISSUE THREAD

**Issue by ree1261** [sh1nobuu/BitAnime#17]
I believe something for gogoanime was changed, but now the downloads are met with a 403 error. When anything is searched, the program displays 404 to the user , but the error is actually 403. I've been struggling with using python to download from gogoanime as well and have also been running into the same errors. Currently, I haven't found any solutions yet.

**Comment by sh1nobuu** [comment]
> I believe something for gogoanime was changed, but now the downloads are met with a 403 error. When anything is searched, the program displays 404 to the user , but the error is actually 403. I've been struggling with using python to download from gogoanime as well and have also been running into the same errors. Currently, I haven't found any solutions yet.

Are you using the .exe file by any chance? I also created a branch (added server) https://github.com/sh1nobuu/BitAnime/tree/feature/add-new-server can't merge it with the master branch yet since the server 3 still not functional.

**Comment by ree1261** [comment]
Just double checked; it was the .exe file. Version 2.0.0. I've given up on trying to scrape gogoanime because I could not for the life of me figure out how to bypass the 403 error. I'm not very experienced, but if you want, I could discuss with you what I know.

**Comment by sh1nobuu** [comment]
> Just double checked; it was the .exe file. Version 2.0.0. I've given up on trying to scrape gogoanime because I could not for the life of me figure out how to bypass the 403 error. I'm not very experienced, but if you want, I could discuss with you what I know.

as of the moment all .exe files are broken since gogoanime decided to change things on their site (again lmao, maybe it was to avoid getting scrape.) about the 403 error if you are talking about when you try to download episodes and you'll get 403 instead the problem is on the header if still trying to find a way to bypass that error. I still don't know how the header on that thing works. https://github.com/sh1nobuu/BitAnime/tree/feature/add-new-server the streamsb on this branch is functional you can use it to download episodes.

**Comment by ree1261** [comment]
I heard it has something to do with the headers as well. Although I'm curious, where and how did you get stream sb links?

**Comment by sh1nobuu** [comment]
> I heard it has something to do with the headers as well. Although I'm curious, where and how did you get stream sb links?

It's really easy. as you can see there's a direct download link and a mirror link after doing some research (?) I found out that the direct links are the unstable one so if i can use the mirror link to download episodes it will solve some of the problem the app have.
![image](https://user-images.githubusercontent.com/73153525/148485051-84e15cb5-2d58-40ec-8e27-4d0d8125594a.png)

**Comment by ree1261** [comment]
The reason I avoided the mirror links was because they required too much clicking. I didn't realize that their download urls followed a pattern. In your code, you were able to get the id, mode, and hash for StreamSB. How did you do that? This is really great because if I don't need gogo if I can scrape sb. 

Also, it looks like gogo changed something again because your new branch fails in startup. I tried the same code a few days ago, and while it said the sb server was available, it wouldn't download for me. I wanted to further investigate today, but it crashed.

Yikes, this whole gogo thing is a nightmare.

**Comment by sh1nobuu** [comment]
> The reason I avoided the mirror links was because they required too much clicking. I didn't realize that their download urls followed a pattern. In your code, you were able to get the id, mode, and hash for StreamSB. How did you do that? This is really great because if I don't need gogo if I can scrape sb.
> 
> Also, it looks like gogo changed something again because your new branch fails in startup. I tried the same code a few days ago, and while it said the sb server was available, it wouldn't download for me. I wanted to further investigate today, but it crashed.
> 
> Yikes, this whole gogo thing is a nightmare.

Yep they do required a lot of clicking, but inspect element is your friend if you can figure out how the link/url generation work you can do it. Oh it's broken again? Thanks for the feedback I will try it on my end.

**Comment by karl0ss** [comment]
I have forked this repo, and have managed to get this all working again, although I have used a bit of your original code, I have made a number of changes that would change BitAnime too much, so I have rebranded to GoGo Downloader -

https://github.com/karl0ss/GoGoDownloader

All original basic functions are working again, I do need to do some work to add some of the nice bits that you did have, thought I would let you know, if you didn't want to spend the time on BitAnime

**Comment by FireHead90544** [comment]
I have rewritten a whole new anime downloader based on GogoAnime. Though I've coded everything from scratch but this repository is one of the main inspiration for doing so and yes a bit of README template is taken from this repo too :)
Here is the link to it, named it "SenPY":

https://github.com/FireHead90544/SenPY

It is much faster since I've used aria2 as external downloader, the logic is pretty much same but sessioning with login cookies is done.
